explorer/ast/value.h

```diff
@@ -95,6 +96,38 @@ class Value {
   const Kind kind_;
 };
 
+// Contains the result of the evaluation of an expression, including a value,
+// the original expression category, and an optional address if available.
+class ExpressionResult {
+ public:
+  ExpressionResult(Nonnull<const Value*> v, std::optional<Address> address,
+                   ExpressionCategory cat)
+      : value_(v), address_(std::move(address)), expr_cat_(cat) {}
+
+  static auto Value(Nonnull<const Carbon::Value*> v) -> ExpressionResult {
+    return ExpressionResult(v, std::nullopt, ExpressionCategory::Value);
+  }
+  static auto Reference(Nonnull<const Carbon::Value*> v, Address address)
+      -> ExpressionResult {
+    return ExpressionResult(v, std::move(address),
+                            ExpressionCategory::Reference);
+  }
+  static auto Initializing(Nonnull<const Carbon::Value*> v, Address address)
+      -> ExpressionResult {
+    return ExpressionResult(v, std::move(address),
+                            ExpressionCategory::Initializing);
+  }
+
+  auto value() const -> Nonnull<const Carbon::Value*> { return value_; }
+  auto address() const -> std::optional<Address> { return address_; }
```

Comment:
```suggestion
  auto address() const -> const std::optional<Address>& { return address_; }
```

Not sure if you just missed this here, but the Address is expensive to copy.

Reply:
Definitely a miss, done.